csrc/multidevice/communication.cpp

```diff
@@ -70,18 +72,48 @@ void assertBuffersHaveSameSize(
   const auto shape = (bufs1.empty() ? bufs2 : bufs1).at(0).sizes();
   for (const auto& bufs : {bufs1, bufs2}) {
     for (const auto& buf : bufs) {
+      NVF_ERROR_EQ(buf.sizes(), shape, "all buffers must have the same shape");
+    }
+  }
+}
+
+// View `t` as a compact tensor that's 1D and contiguous. All non-broadcast
+// dimensions of `t` must be contiguous. Note that this definition is different
+// from torch.Tensor.is_contiguous in that torch.Tensor.is_contiguous doesn't
+// allow expanded broadcast dimensions (aka internal overlapping).
+at::Tensor viewAsCompact(at::Tensor t) {
+  int64_t n_elements = t.numel();
+  for (auto [size, stride] : zip(t.sizes(), t.strides())) {
+    if (stride == 0) {
       NVF_ERROR(
-          buf.sizes() == shape,
-          "all buffers must have the same shape, but got: ",
-          buf.sizes(),
-          " vs ",
-          shape);
+          n_elements % size == 0,
+          "n_bytes is expected to be divisible by size, but got ",
+          n_elements,
+          " and ",
+          size);
+      n_elements /= size;
     }
   }
+  return t.as_strided(/*sizes=*/{n_elements}, /*strides=*/{1});
 }
 
-void doLocalCopy(const at::Tensor& dst, const at::Tensor& src) {
-  dst.view_as(src).copy_(src, /*non_blocking=*/true);
+// Copy `src` to `dst`. `src` and `dst` must satisfy `isTvContiguous`.
+void doLocalCopy(at::Tensor dst, at::Tensor src) {
+  NVF_ERROR_EQ(dst.numel(), src.numel());
+  if (dst.is_contiguous()) {
+    dst.view_as(src).copy_(src, /*non_blocking=*/true);
+    return;
+  }
+
+  // `isTVContiguous` allows tensors with expanded broadcasts, which fail
+  // `torch.Tensor.is_contiguous`.
+  at::Tensor dst_compact = viewAsCompact(dst);
```

Comment:
I believe `copy_` in the above case uses `cudaMemcpyAsync`: https://github.com/pytorch/pytorch/blob/56218d85e2da09d9ede3809718ec989c2151632c/aten/src/ATen/native/cuda/Copy.cu#L403.

So we should we able to directly just use that and remove the `if` earlier.

Reply:
> I believe copy_ in the above case uses cudaMemcpyAsync

Thanks for figuring this out! 

> we should be able to directly just use that and remove the if earlier.

Yes, I'm happy to do that.